x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,847 @@
+            final MetadataIdentifier metadataIdentifier = metadataIdentifierSupplier.get();
+            final String metadataBlobName = MetadataIdentifier.formMetadataBlobName(blobName, metadataIdentifier);
+            // first write the encrypted metadata to a UNIQUE blob name
+            try (ByteArrayInputStream encryptedMetadataInputStream = new ByteArrayInputStream(encryptedMetadata)) {
+                encryptionMetadataBlobContainer.writeBlob(metadataBlobName, encryptedMetadataInputStream, encryptedMetadata.length, true
+                        /* fail in the exceptional case of metadata blob name conflict */);
+            }
+            // afterwards write the encrypted data blob
+            // prepended to the encrypted data blob is the unique identifier (fixed length) of the metadata blob
+            final long encryptedBlobSize = (long) MetadataIdentifier.byteLength() +
+                    EncryptionPacketsInputStream.getEncryptionLength(blobSize, PACKET_LENGTH_IN_BYTES);
+            try (InputStream encryptedInputStream =
+                         ChainingInputStream.chain(new ByteArrayInputStream(metadataIdentifier.asByteArray()),
+                                 new EncryptionPacketsInputStream(inputStream, dataEncryptionKey, nonce, PACKET_LENGTH_IN_BYTES))) {
+                delegatedBlobContainer.writeBlob(blobName, encryptedInputStream, encryptedBlobSize, failIfAlreadyExists);
+            }
+        }
+
+        @Override
+        public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+            // the encrypted repository does not offer an alternative implementation for atomic writes
+            // fallback to regular write
+            writeBlob(blobName, inputStream, blobSize, failIfAlreadyExists);
+        }
+
+        @Override
+        public DeleteResult delete() throws IOException {
+            // first delete the encrypted data blob
+            DeleteResult deleteResult = delegatedBlobContainer.delete();
+            // then delete metadata
+            try {
+                deleteResult = deleteResult.add(encryptionMetadataBlobContainer.delete());
+            } catch (IOException e) {
+                // the encryption metadata blob container might not exist at all
+                logger.warn("Failure to delete metadata blob container " + encryptionMetadataBlobContainer.path(), e);
+            }
+            return deleteResult;
+        }
+
+        @Override
+        public void deleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException {
+            Objects.requireNonNull(blobNames);
+
+            // find all the blob names that must be deleted
+            Set<String> blobNamesSet = new HashSet<>(blobNames);
+            Set<String> blobNamesToDelete = new HashSet<>();
+            for (String existingBlobName : delegatedBlobContainer.listBlobs().keySet()) {
```

Comment:
This we shouldn't do. We should just optimistically delete the blobs and assume they're all there. It's really costly to run this listing and it may not be accurate on S3 and would cause you to not try and delete a certain blob if it doesn't show up in the listing I think?

Reply:
> We should just optimistically delete the blobs and assume they're all there.

There is a problem with this approach:
For every existing data blob there must be a metadata blob already visible, because metadata blobs are written first. For every blob name, you must delete the meta and the data blobs. But then because the meta blobs are written first, it's possible that a `deleteBlob` operation will delete just the meta and fail to delete the data because that's not visible yet, and the data blob might appear later.

This approach makes it sure that you only delete the meta of existing data blobs, so that no corruption is possible. Am I right?

> it may not be accurate on S3 and would cause you to not try and delete a certain blob if it doesn't show up in the listing

I believe that the limitation of deleting only the blobs you can list is OK.

Let's discuss, maybe we can come up with a more performant alternative.